isabelle: make closer to upstream

sledgehammervampire opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Incorporates some changes from #115339.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jvanbruegge commented on pkgs/top-level/all-packages.nix:
```diff
@@ -32416,8 +32416,16 @@ with pkgs;
       configureFlags = [ "--enable-intinf-as-int" "--with-gmp" "--disable-shared" ];
+      buildFlags = [ "compiler" ];
+      version = "for-isabelle-2021-1";
+      src = fetchFromGitHub {
```
Any particular reason to lock the polyml version here?

1000teslas replied:
I want to use a version that matches the upstream distribution as closely as possible just in case Isabelle relies on it. I know that Isabelle relies on a particular version of Z3, and using the version from nixpkgs has broken things before (#99399).

jvanbruegge replied:
I see

jvanbruegge commented:
A few of my PRs were merged, so you would have to rebase this

jvanbruegge commented:
@gebner @SuperSandro2000 this would be ready to merge as well

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -32765,8 +32765,16 @@ with pkgs;
   isabelle = callPackage ../applications/science/logic/isabelle {
-    polyml = lib.overrideDerivation polyml (_: {
+    polyml = polyml.overrideAttrs (_: {
       configureFlags = [ "--enable-intinf-as-int" "--with-gmp" "--disable-shared" ];
+      buildFlags = [ "compiler" ];
+      version = "for-isabelle-2021-1";
```
```suggestion
      pname = "polyml-for-isabelle";
      version = "2021-1";
      configureFlags = [ "--enable-intinf-as-int" "--with-gmp" "--disable-shared" ];
      buildFlags = [ "compiler" ];
```

SuperSandro2000 commented on pkgs/applications/science/logic/isabelle/default.nix:
```diff
@@ -1,6 +1,38 @@
+  };
+  ml_home = runCommand "ml-home" { } ''
+    mkdir $out
+    ln -s ${polyml}/bin/* $out
+    ln -s ${polyml}/lib/* $out
+    ln -s ${sha1}/lib/* $out
```
Does this even end up in someones configuration? If so we can't link thing to /.

1000teslas replied:
Sorry, I don't understand what you mean by "can't link thing to /".

jvanbruegge replied:
@SuperSandro2000 could you clarify what you mean by this?

jvanbruegge replied:
I think he meant that libraries should go to `$out/lib` and binaries to `$out/bin`. Could you change that and see if it still works?

1000teslas replied:
Oh, I don't think that will work, since I'm pretty sure Isabelle wants libsha1.so in the same directory as the poly binary.

jvanbruegge replied:
I have a patch that fixes that. Is it ok if I take over here? I won't be able to push to your branch, so I would have to open a new PR

jvanbruegge replied:
If you want to amend your PR, here is the code: https://github.com/jvanbruegge/nixpkgs/commit/9aa4c5dbfc806b2f71ad351fc79d5d0312e0caa4

1000teslas replied:
I've amended my PR.

SuperSandro2000 commented on pkgs/applications/science/logic/isabelle/default.nix:
```diff
@@ -1,6 +1,38 @@
 
+let
+  sha1 = stdenv.mkDerivation {
+    name = "isabelle-sha1";
```
Please use pname+version.

1000teslas replied:
Not sure what version to put, since I'm fetching from a specific git commit.

1000teslas replied:
I've decided to specify the version as `2021-1`.

jvanbruegge committed:
isabelle: Make closer to upstream

This makes isabelle use the versions of polyml and z3 that are also used
upstream. Additionally it packages the sha1 library that isabelle uses.

Co-authored-by: 1000teslas <47207223+1000teslas@users.noreply.github.com>

jvanbruegge commented:
@SuperSandro2000 is this PR good to go now?
